positive negative (#188)
andycroll/andycroll.com: _posts/2023/2023-09-11-express-yourself-clearly-with-positive-negative-numbers.md

---
title: "Express yourself clearly with positive? and negative? for numbers"
description: "Syntactical sugar to avoid comparison operators"
layout: article
category: ruby
image:
  base: "2023/express-yourself-clearly-with-positive-negative-numbers"
  alt: "The architecture of Birmingham, UK"
  credit: "Mike Hindle"
  source: "https://unsplash.com/photos/ow87XYsaY6Q"

---

> Ruby, in contrast to other languages, often provides multiple ways to accomplish simple programming tasks. In pursuit of developer happiness the Standard Library offers the opportunity to make your code closer to English.

Reviewer: ```suggestion
Ruby, in contrast to other languages, often provides multiple ways to accomplish simple programming tasks. In pursuit of developer happiness, the Standard Library offers the opportunity to make your code appear more like English.
```

Reviewer: Not 100% convinced on my suggested change, but "closer to English" felt like it needed something!

> In this case we’ll look at comparing numbers, I’ve previously discussed my preference for a similar comparator, `zero?`, in a [previous article](/ruby/use-zero-for-integer-float-comparison).

Reviewer: ```suggestion
In this case we’ll look at comparing numbers. I’ve previously discussed my preference for a similar comparator, `zero?`, in a [previous article](/ruby/use-zero-for-integer-float-comparison).
```

I feel like this should go after you've shown us what you're showing us.

## Instead of…

…using comparison operators:

```ruby
if number > 0
  # do a thing
end

if number < 0
  # do a thing
end
```

## Use…

…ruby’s convenience methods on `Numeric`, and its subclasses, to express yourself more clearly:

```ruby
if number.positive?
  # do a thing
end

if number.negative?
  # do a thing
end
```

## Why?

> I find the “more English” version to be clearer and easier to reason about when I return to it in the future. I still, after a twenty year programming career, require a moment of concentration while I unpack the orientation of the angle bracket and how it relates to the numbers and variables.

Reviewer: Less "I" please. More general, second-person language. Maybe the bit where you figure out which direction the bracket implies can still be first-person.

> This approach to convenience methods also, for me, is one of the most delightful things about Ruby. It’s one of the reasons it provokes smiles when I write it.

Reviewer: Again, would make this less "for me/I" and a little more general.

</tr>
<tr>
  <th>-1 > 0</th>
  <td class="text-right">30.919M <small>(± 0.1%)</small> i/s</td>
</tr>
<tr>
  <th>1.positive?</th>
  <td class="text-right">22.602M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>0.positive?</th>
  <td class="text-right">22.621M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>-1.0.positive?</th>
  <td class="text-right">22.676M <small>(± 0.1%)</small> i/s</td>
</tr>
</table>

For `Float`:

<table>
<tr>
  <th>1.0 > 0</th>
  <td class="text-right">31.442M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>0.0 > 0</th>
  <td class="text-right">31.178M <small>(± 0.1%)</small> i/s</td>
</tr>
<tr>
  <th>-1.0 > 0</th>
  <td class="text-right">30.861M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>1.0.positive?</th>
  <td class="text-right">22.707M <small>(± 0.1%)</small> i/s</td>
</tr>
<tr>
  <th>0.0.positive?</th>
  <td class="text-right">22.576M <small>(± 0.1%)</small> i/s</td>
</tr>
<tr>
  <th>-1.0.positive?</th>
  <td class="text-right">22.654M <small>(± 0.1%)</small> i/s</td>
</tr>
</table>

The benchmark shows that both for `Integer` and `Float`s the _non-sugared_ syntax wins out.

> However, the important thing to note is that in all cases I was achieving tens of millions of executions per second on my laptop, for anything other than the most extreme performance requirements I’d prefer the more readable version. And if performance _truly_ matters beyond this scale then you likely have bigger issues!

Reviewer: ```suggestion
However, the important thing to note is that in all cases I was achieving tens of millions of executions per second on my laptop. For anything other than the most extreme performance requirements opt for the more readable version. And if performance _truly_ matters beyond this scale, then you likely have bigger issues!
```

---
title: "Express yourself clearly with positive? and negative? for numbers"
description: "Syntactical sugar to avoid comparison operators"
layout: article
category: ruby
image:
  base: "2023/express-yourself-clearly-with-positive-negative-numbers"
  alt: "The architecture of Birmingham, UK"
  credit: "Mike Hindle"
  source: "https://unsplash.com/photos/ow87XYsaY6Q"

---

Ruby, in contrast to other languages, often provides multiple ways to accomplish simple programming tasks. In pursuit of developer happiness, the Standard Library offers the opportunity to make your code appear more like English.

In this case we’ll look at comparing numbers with zero.

## Instead of…

…using comparison operators:

```ruby
if number > 0
  # do a thing
end

if number < 0
  # do a thing
end
```

## Use…

> …ruby’s convenience methods on `Numeric`, and its subclasses, to express yourself more clearly:

Reviewer: *Ruby's

Author: ```suggestion
…Ruby’s convenience methods on `Numeric`, and its subclasses, to express yourself more clearly:
```

```ruby
if number.positive?
  # do a thing
end

if number.negative?
  # do a thing
end
```

## Why?

> I’ve previously discussed my preference for a similar comparator, `zero?`, in a [previous article](/ruby/use-zero-for-integer-float-comparison).

Reviewer: ```suggestion
I’ve discussed my preference for a similar comparator, `zero?`, in a [previous article](/ruby/use-zero-for-integer-float-comparison).
```

> The “more English” version to be clearer and easier to reason about when you return to your code in the future. It can require a moment of concentration to unpack the orientation of the angle bracket and how it relates to the numbers and variables.

Reviewer: *will be clearer

Author: ```suggestion
The “more English” version will be clearer and easier to reason about when you return to your code in the future. It can require a moment of concentration to unpack the orientation of the angle bracket and how it relates to the numbers and variables.
```

> This approach to convenience methods and syntactical sugar is one of the most delightful things about Ruby. “Ruby is designed to make programmers happy.”–Yukihiro “Matz” Matsumoto

Reviewer: New line for the quote?

Author: ```suggestion
This approach to convenience methods and syntactical sugar is one of the most delightful things about Ruby.

“Ruby is designed to make programmers happy.”–Yukihiro “Matz” Matsumoto
```

## Why not?

Some folks find this style to be too different from other languages and claim that they’ve never struggled to remember whether `<` is “less than” or “greater than’.

> You might get push back for “performance” reasons.

Reviewer: I'd do a colon at the end of this sentence.

Author: ```suggestion
You might get push back for “performance” reasons:
```

</tr>
<tr>
  <th>0 > 0</th>
  <td class="text-right">31.625M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>-1 > 0</th>
  <td class="text-right">30.919M <small>(± 0.1%)</small> i/s</td>
</tr>
<tr>
  <th>1.positive?</th>
  <td class="text-right">22.602M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>0.positive?</th>
  <td class="text-right">22.621M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>-1.0.positive?</th>
  <td class="text-right">22.676M <small>(± 0.1%)</small> i/s</td>
</tr>
</table>

For `Float`:

<table>
<tr>
  <th>1.0 > 0</th>
  <td class="text-right">31.442M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>0.0 > 0</th>
  <td class="text-right">31.178M <small>(± 0.1%)</small> i/s</td>
</tr>
<tr>
  <th>-1.0 > 0</th>
  <td class="text-right">30.861M <small>(± 0.2%)</small> i/s</td>
</tr>
<tr>
  <th>1.0.positive?</th>
  <td class="text-right">22.707M <small>(± 0.1%)</small> i/s</td>
</tr>
<tr>
  <th>0.0.positive?</th>
  <td class="text-right">22.576M <small>(± 0.1%)</small> i/s</td>
</tr>
<tr>
  <th>-1.0.positive?</th>
  <td class="text-right">22.654M <small>(± 0.1%)</small> i/s</td>
</tr>
</table>

> The benchmark shows that both for `Integer` and `Float`s the _non-sugared_ syntax wins out.

Reviewer: ```suggestion
The benchmark shows that, for both `Integer` and `Float`s, the _non-sugared_ syntax wins out.
```

> However, the important thing to note is that in all cases I was achieving tens of millions of executions per second on my laptop. For anything other than the most extreme performance requirements opt for the more readable version. And if performance _truly_ matters beyond this scale, then you likely have bigger issues!

Reviewer: Comma after "requirements" ?

Author: ```suggestion
However, the important thing to note is that in all cases I was achieving tens of millions of executions per second on my laptop. For anything other than the most extreme performance requirements, opt for the more readable version. And if performance _truly_ matters beyond this scale, then you likely have bigger issues!
```
